doc/watermarks_HLD.md

```diff
@@ -0,0 +1,352 @@
+| COUNTERS_DB | "HIGHEST_WATERMARKS/Ethernet``<port number``>/Queues" | Queue highest  watermarks on one Ethernet ports |
+| COUNTERS_DB | "HIGHEST_WATERMARKS/Ethernet*/PriorityGroups" | PG highest watermarks on all Ethernet ports |
+| COUNTERS_DB | "HIGHEST_WATERMARKS/Ethernet``<port number``>/PriorityGroups" | PG highest watermarks on one Ethernet ports |
+| COUNTERS_DB | "WATERMARKS/Ethernet*/PriorityGroups/``<watermark name``>" | Specific PG watermark on all Ethernet ports |
+| COUNTERS_DB | "HIGHEST_WATERMARKS/Ethernet``<port number``>/PriorityGroups/``<watermark name``>" | Specific PG highest watermark on one Ethernet ports |
+
+
+### 4 Flows
+
+#### 4.1 Watermark general flow
+
+![](https://github.com/mykolaf/SONiC/blob/gh-pages/images/watermark_HLD/WM_general.png)
+
+The core components are the flex counter, watermark orch, DB, CLI. 
+
+The flex counter reads and clears the watermarks on a peroid of 1s by default. The values are put directly to COUNTERS table. The flex counter also has plugins configured for queue and pg, which will be triggered on every flex counter group interval. The lua plugin will update WATERMARKS, HIGHEST_WATERMARKS and PERIOD_WATERMARKS with if the new value exceeds the vlaue that was read from the table.  
+
+The watermark orch has 2 main functions:
+ - Handle the Timer that clears the WATERMARKS table. Handle the configuring of the interval for the timer.
+ - Handle Clear notificatons. On clear event the orch should just zero-out the corresponding watermarks from the table. It will be soon repopulated by lua plugin.  
+
+The DB contains all the tables with watemarks, and the configuration table.
+
+The Cli reads the watermarks from the tables, formats and outputs it. For the watermark over period scenario, it also sets a field in the config table to instruct lua script to also track PERIOD_WATERMARK,
+sleeps for the period, then gets the watermarks from PERIOD_WATERMARKS table(see flow).
+
+#### 4.2 Resetting the telemetry period flow
+
+![](https://github.com/mykolaf/SONiC/blob/gh-pages/images/watermark_HLD/WM_period.PNG)
+
+The watermark orch handles notifications on changes in WATERMARK_TABLE in config DB. The new interval will be assigned to the timer during the timer handling, so the orch will reset the interval only when the current timer expires.
+
+#### 4.3 Cli flow
+
+![](https://github.com/mykolaf/SONiC/blob/gh-pages/images/watermark_HLD/WM_cli.png)
+
+### 5 Open questions
+
+#### Should I worry much about multiple regular users interfering with each other?
+
+This is a debug feature aimed for only one user and the requirements say "we consider a maximum of one regular user" so I disregard this possibility for now. 
+
+#### Should I consider a scenario in which a user is querrying a watermark over a period of time for multiple types(f. e. pg headroom and pg shared) of watermarks at the same time?
+
```

Comment:
depending the performance, we may need to increase the polling interval for the watermark and introduce on-demand asic counter clear on the watermarks.